lib/rules/unit-disallowed-list/index.js

```diff
@@ -96,75 +113,109 @@ const rule = (primary, secondaryOptions) => {
-					// the values not included in the parentValue string
-					if (valueNode.type !== 'word' || !parentValue.includes(valueNode.value)) {
-						return;
+		root.walkAtRules(/^media$/i, (atRule) => {
+			const params = getAtRuleParams(atRule);
+
+			if (!hasDimension(params)) return;
+
+			parseFromTokens(tokenizeWithoutPercentageTokens(params)).forEach((mediaQuery) => {
+				/** @type {{ mediaFeatureName: string | undefined }} */
+				const state = {
+					mediaFeatureName: undefined,
+				};
+
+				mediaQuery.walk((entry) => {
+					if (!entry.state) return;
+
+					if (isMediaFeature(entry.node)) {
+						entry.state.mediaFeatureName = entry.node.getName().toLowerCase();
 					}
 
+					if (!entry.state.mediaFeatureName) return;
+
+					if (!isTokenNode(entry.node)) return;
+
 					check(
-						node,
-						getIndex(node),
-						valueNode,
-						mediaName,
-						secondaryOptions ? secondaryOptions.ignoreMediaFeatureNames : {},
+						atRule,
+						atRuleParamIndex,
+						entry.node,
+						entry.state.mediaFeatureName,
+						secondaryOptions?.ignoreMediaFeatureNames,
 					);
-				});
+				}, state);
 			});
-		}
+		});
+		root.walkDecls((decl) => {
+			const value = getDeclarationValue(decl);
+
+			if (!hasDimension(value)) return;
+
+			parseCommaSeparatedListOfComponentValues(tokenize({ css: value }))
+				.flatMap((x) => x)
+				.forEach((componentValue) => {
+					if (isTokenNode(componentValue)) {
+						check(
+							decl,
+							declarationValueIndex,
+							componentValue,
+							decl.prop,
+							secondaryOptions?.ignoreProperties,
+						);
 
-		/**
-		 * @template {import('postcss').Declaration} T
-		 * @param {T} node
-		 * @param {string} value
-		 * @param {(node: T) => number} getIndex
-		 * @returns {void}
-		 */
-		function checkDecl(node, value, getIndex) {
-			// make sure multiplication operations (*) are divided - not handled
-			// by postcss-value-parser
-			value = value.replace(/\*/g, ',');
-
-			valueParser(value).walk((valueNode) => {
-				const valueLowerCase = valueNode.value.toLowerCase();
-
-				// Ignore wrong units within `url` function
-				if (
-					valueNode.type === 'function' &&
-					(valueLowerCase === 'url' ||
-						optionsMatches(secondaryOptions, 'ignoreFunctions', valueLowerCase))
-				) {
-					return false;
-				}
-
-				check(
-					node,
-					getIndex(node),
-					valueNode,
-					node.prop,
-					secondaryOptions ? secondaryOptions.ignoreProperties : {},
-				);
-			});
-		}
+						return;
+					}
 
-		root.walkAtRules(/^media$/i, (atRule) => checkMedia(atRule, atRule.params, atRuleParamIndex));
-		root.walkDecls((decl) => checkDecl(decl, decl.value, declarationValueIndex));
+					if (!isFunctionNode(componentValue) && !isSimpleBlockNode(componentValue)) return;
+
+					componentValue.walk(
+						(entry) => {
+							if (!entry.state) return;
+
+							if (entry.state.ignored) return;
+
+							if (isTokenNode(entry.node) && !entry.state.ignored) {
```

Comment:
[ask] Does this `!entry.state.ignored` duplicate `if (entry.state.ignored) return;` at L174? Can we remove the check?

```suggestion
							if (isTokenNode(entry.node)) {
```